### PR TITLE
Ir generation

### PR DESCRIPTION
**Pull request by boki1**
### Description

Provide an implementation of intermediate code generation by having an already constructed AST.

### Task

Using additional context storage implement a translation (compilation) unit which is obligated with the task of traversing the AST in a way that enables the artificial instructions to be produced.

### Definition Of Done

Successfully translating a given abstract syntax tree to a set of “fake” instructions (three-address code).

### CODE REVIEW

File: src/frontend/ir_generation/codes.cc [hunk 1]
```diff
@@ -0,0 +1,84 @@
+/**
```
Comment (boki1) [inline src/frontend/ir_generation/codes.cc:1]: There isn't very detailed documentation. Maybe some clarifications above the function declarations might help.

File: tools/clean.sh [hunk 2]
```diff
@@ -2,7 +2,10 @@ rm -rf build
...
 rm -rf Testing
 rm -rf Debug
-rm -r frontend/parser/*tab*
-rm -r frontend/parser/position.hh
-rm -r frontend/parser/location.hh
-rm -r frontend/parser/stack.hh
+
+PARSER_DIR=src/frontend/parser
+
+rm -f $PARSER_DIR/*tab*
```
Comment (vkabadzhova) [inline tools/clean.sh:8]: Good addition :wink:

File: src/frontend/ir_generation/concrete_generators.cc [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    void gu::generate_sequence(const syntax_analyzer::expression &e, generation_context &gtx) {
+        for (auto &i : e.operands()) {
+            tac::vregister_type prev = gtx.pop_temp();
+            tac::vregister_type last = generate_ir(i, gtx);
+            gtx.push_temp(last);
+
+            if (prev == ~0u)
+                continue;
+
+            tac *(generation_unit::*fun)(const tac::operands_type &operands);
```
Comment (boki1) [inline src/frontend/ir_generation/concrete_generators.cc:114]: I really like that approach :smile: :+1:

File: src/frontend/ir_generation/generation_unit.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		}
+
+		case expr_type::Addition:
+		case expr_type::Equality:
+		case expr_type::Sequence:
+		case expr_type::Multiplication:
+		case expr_type::ModularDivision:
+		case expr_type::Division:
+		{
+			// TODO:
```
Comment (boki1) [inline src/frontend/ir_generation/generation_unit.cc:62]: Is this comment actually still relevant?

File: src/frontend/ir_generation/generation_unit.hh [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		void generate_identifier(const syntax_analyzer::expression& e, generation_context& gtx);
+
+		void generate_conditional(const syntax_analyzer::expression& expr, generation_context& gtx);
+
+	 public: // TAC concrete constructors (codes.cc)
+		[[nodiscard]] tac* tac_nop();
+		[[nodiscard]] tac* tac_init(tac::vregister_type value, const std::string& ident, unsigned operand);
+		[[nodiscard]] tac* tac_add(const tac::operands_type& operands);
+		[[nodiscard]] tac* tac_neg(const tac::operands_type& operands);
+        [[nodiscard]] tac* tac_mult(const tac::operands_type& operands);
```
Comment (boki1) [inline src/frontend/ir_generation/generation_unit.hh:73]: This indentation is kind of weird :smile:. Maybe that could be improved.

File: src/frontend/ir_generation/generation_unit.hh [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+	 public:
+
+		/// Instantiates a context for the given function and returns it
+		generation_unit::generation_context
+		prepare_context(const syntax_analyzer::function& function);
+
+	 public: // Concrete generators (concrete_generators.cc)
+
+		void generate_sequence(const syntax_analyzer::expression& e, generation_context& gtx);
+
```
Comment (boki1) [inline src/frontend/ir_generation/generation_unit.hh:57]: Maybe some doc-comments :smile: :memo: :speech_balloon:

File: src/frontend/ir_generation/ir_code.hh [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+		/**
+		 * @brief Enumeration class containing all the three-address codes
+		 */
+		enum class type
+		{
+			Nop,            ///< No operation
+			Init,           ///< Store a pointer to a value at address &identifier + offset where identifier is a function's base
+			Add,            ///< Perform addition
+			Negate,         ///< Negation
+            Multiplication, ///< Multiplication
```
Comment (boki1) [inline src/frontend/ir_generation/ir_code.hh]: The indentation problem again :smiley:

File: src/frontend/ir_generation/ir_code.hh [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            Division,       ///< Division
+            ModuloOperator,         ///< ModuloOperator operator
+			Copy,           ///< Assign a copy of another variable
+			Read,           ///< Read from address
```
Comment (boki1) [inline src/frontend/ir_generation/ir_code.hh]: Didn't we mark these as obsolete (`Read` and `Write`) or at least not useful?

File: src/test/frontend/ir_generation.test.cc [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/**
```
Comment (boki1) [inline src/test/frontend/ir_generation.test.cc:1]: Maybe some additions to this file :smiley: ?

**Review by vkabadzhova — APPROVED** [review]

**Pull request merged**